runtime/memprof.c

```diff
@@ -451,48 +592,54 @@ void caml_memprof_track_young(tag_t tag, uintnat wosize, int from_caml)
   Caml_state->young_ptr += whsize;
   caml_memprof_renew_minor_sample();
 
-  /* Empty the queue to make sure callbacks are called in the right
-     order. */
-  caml_raise_if_exception(caml_memprof_handle_postponed_exn());
-
+  caml_memprof_suspended = 1;
   callstack = capture_callstack();
-  ephe = caml_raise_if_exception(do_callback_exn(tag, wosize, occurrences,
-                                                 callstack, Minor));
+  t_idx = new_tracked(occurrences, Make_header(wosize, tag, Caml_white),
+                      0, 1, Placeholder_value, callstack);
+  res = (t_idx == Invalid_index) ? run_any_callback_exn(&t_idx) : Val_unit;
+  caml_memprof_suspended = 0;
+  caml_memprof_check_action_pending();
 
   /* We can now restore the minor heap in the state needed by
      [Alloc_small_aux]. */
-  if (Caml_state->young_ptr - whsize < Caml_state->young_trigger) {
-    CAML_INSTR_INT ("force_minor/memprof@", 1);
+  if(Caml_state->young_ptr - whsize < Caml_state->young_trigger) {
+    CAML_INSTR_INT("force_minor/memprof@", 1);
     caml_gc_dispatch();
   }
 
+  caml_raise_if_exception(res);
+
   /* Re-allocate the block in the minor heap. We should not call the
      GC after this. */
   Caml_state->young_ptr -= whsize;
 
   /* Make sure this block is not going to be sampled again. */
   shift_sample(whsize);
 
-  /* Write the ephemeron if not [None]. */
-  if (Is_block(ephe)) {
-    /* Subtlety: we are actually writing the ephemeron with an invalid
-       (uninitialized) block. This is correct for two reasons:
-          - The logic of [caml_ephemeron_set_key] never inspects the content of
-            the block. In only checks that the block is young.
-          - The allocation and initialization happens right after returning
-            from [caml_memprof_track_young]. */
-    caml_ephemeron_set_key(Field(ephe, 0), 0, Val_hp(Caml_state->young_ptr));
+  if (t_idx == Invalid_index)
```

Comment:
This condition happens in other cases than OOM: `t_idx` has been set by the callback, which could invalidate `t_idx`.